gcloud/storage/batch.py

```diff
@@ -0,0 +1,209 @@
+            buf = io.StringIO()
+        else:                   # pragma: NO COVER  Python2
+            buf = io.BytesIO()
+        generator = Generator(buf, False, 0)
+        generator.flatten(multi)
+        payload = buf.getvalue()
+
+        # Strip off redundant header text
+        _, body = payload.split('\n\n', 1)
+        headers = dict(multi._headers)
+
+        url = self._connection.build_api_url('/batch')
+
+        _req = self._connection._make_request
+        response, content = _req('POST', url, data=payload, headers=headers)
+        self._responses = list(_unpack_batch_response(response, content))
+        return self._responses
+
+    def __enter__(self):
+        _BATCHES.push(self)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        try:
+            if exc_type is None:
+                self.finish()
+            else:
+                self._http.reset()
+        finally:
+            _BATCHES.pop()
+
+
+def _unpack_batch_response(response, content):
+    """Convert response, content -> [(status, reason, payload)]."""
+    parser = Parser()
+    faux_message = ('Content-Type: %s\nMIME-Version: 1.0\n\n%s' %
+                    (response['Content-Type'], content))
+
+    message = parser.parsestr(faux_message)
+
+    if not isinstance(message._payload, list):
+        raise ValueError('Bad response:  not multi-part')
+
+    for subrequest in message._payload:
+        status_line, rest = subrequest._payload.split('\n', 1)
+        _, status, reason = status_line.split(' ', 2)
+        message = parser.parsestr(rest)
+        payload = message._payload
+        ctype = message['Content-Type']
+        if ctype and ctype.startswith('application/json'):
+            payload = json.loads(payload)
+        yield status, reason, payload
+
+
+class NoContent(object):
+    """Emulate an HTTP '204 No Content' response."""
+    status = 204
+
+
+class _FauxHTTP(object):
+    """Emulate ``connection.http``, but store requests.
+
+    Only allow up to ``_MAX_BATCH_SIZE`` requests to be bathed.
+    """
+    _MAX_BATCH_SIZE = 1000
+
+    def __init__(self, connection):
+        self._connection = connection
+        self._requests = []
+        self._orig_http, connection.http = connection.http, self
+
+    def request(self, method, uri, headers, body):
+        """Emulate / proxy underlying HTTP request.
+
+        - Pass 'GET' requests through.
+
+        - Defer others for later processing
+        """
+        if method == 'GET':
+            _req = self._orig_http.request
+            return _req(method=method, uri=uri, headers=headers, body=body)
+
+        if len(self._requests) >= self._MAX_BATCH_SIZE:
+            self.reset()
+            raise ValueError("Too many deferred requests (max %d)" %
+                             self._MAX_BATCH_SIZE)
+
+        self._requests.append((method, uri, headers, body))
+        return NoContent(), ''
```

Comment:
`NoContent` should likely be imported from `gcloud.exceptions`, but we don't map any of the 200-range response codes there yet.